app/controllers/application_controller.rb

```diff
@@ -1,7 +1,41 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  before_action :set_language
+
+  def set_language
+    I18n.locale =
+      if user_signed_in?
+        current_user.language
+      else
+        determine_language
+      end
+  end
+
+  def default_url_options(*)
+    { locale: I18n.locale == I18n.default_locale ? nil : I18n.locale }
```

Comment:
Mind doing `(I18n.locale == I18n.default_locale if I18n.locale)`, just for a slight readability tweak? Took me a minute to figure this out.